server.go

```diff
@@ -282,12 +282,15 @@ func (s *Server) sendResponse(t transport.ServerTransport, stream *transport.Str
 }
 
 func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc) (err error) {
+	ctx, cancel := context.WithCancel(stream.Context())
```

Comment:
I have slight concern that this introduces unnecessary latency on the fast path when tracing is disabled. We need a way to avoid it. But it is okay to put it into a separate pr.